File: chapter13/code13_02.cpp
#include <iostream>
#include <vector>

using namespace std;
using Graph = vector<vector<int>>;

vector<bool> seen;

void dfs(const Graph &G, int v)
{
    seen[v] = true;

    for (auto next_v : G[v])
    {
        if (seen[next_v])
            continue;
        dfs(G, next_v);
    }
}

int main()
{
    //頂点数 N　辺の数 M
    int N, M;
    cin >> N >> M;

    Graph G(N);
    for (int i = 0; i < M; ++i)
    {
        int a, b;
        cin >> a >> b;
        G[a].push_back(b);
        //無向グラフの場合、以下を追加
        //G[b].push_back(a);
    }

    seen.assign(N, false);
    for (int v = 0; v < N; ++v)
    {
        if (seen[v])
            continue;
        dfs(G, v);
    }
}

File: chapter13/code13_02.py


def dfs(G, s, seen):
    seen[s] = True

    for i in G:
        if i[0] == s:
            if seen[i[1]]:
                continue
            dfs(G, i[1], seen)


N, M = map(int, input().split())

Graph = [map(int, input().split()) for _ in range(M)]

seen = [False]*N


for i in range(N):
    if seen[i]:
        continue
    dfs(Graph, i, seen)
